### PR TITLE
Introduce handling of versioned Boolean algebra expressions

### PR DESCRIPTION
**Pull request by evgenyz**
#### Description:

The `ssg/boolean_expression.py` module is based on *boolean.py* (in the `ext` directory) and *pkg_resources* from setuptools (version specifiers PEP440, PEP508).

Examples and tests are in `tests/unit/ssg-module/test_boolean_expressions.py`.

#### Rationale:

We need it for handling CPE platforms and potentially other things.

### CODE REVIEW

**Comment by vojtapolasek** [comment]
Hello, that looks great. I see that tests on SLE and Debian are failing. It seems that ordering of XML attributes got swapped for some reason on these platforms. I can't reproduce it on Fedora.

**Comment by evgenyz** [comment]
> Hello, that looks great. I see that tests on SLE and Debian are failing. It seems that ordering of XML attributes got swapped for some reason on these platforms. I can't reproduce it on Fedora.

Yeah, that's because of how `xml.dom`s `toprettyxml` works in different versions.

File: tests/unit/ssg-module/test_boolean_expressions.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    exp_s = exp.simplify()
+    assert str(exp_s) == 'anything|(~apple&~banana)'
+
+
+def test_evaluate_simple_boolean_ops(exp_simple):
+    assert exp_simple(**{'oranges': '2.0', 'apple': True, 'pie': True})
+    assert not exp_simple(**{'oranges': '2.0', 'apple': False, 'pie': True})
+
+
+def test_evaluate_simple_version_ops(exp_simple):
```
Comment (matejak) [inline tests/unit/ssg-module/test_boolean_expressions.py:57]: I would copy-paste the expression definition, as having it close to the actual test is convenient.

**Comment by evgenyz** [comment]
>     1. The result of parsing the expression is either `Function` or `Symbol`. `Function` is basically a graph node, its `args` are other nodes or leaf nodes. Therefore, the representation of a boolean expression is available in form of a tree.

Yes.
 
>     2. There are other files included - is it about dealing with version numbers?

Yes, the `packaging` module is related to version specifiers.

>     3. There is expression simplification included "for free" - we have to make it clear whether we want it in the full scale though. I remember that we also liked the idea of having platform definitions close to the form in which the platform was defined originally, so the generated content is somewhat recognizable to content creators. That feature comes with its own problems too, so we just have to decide which path to take.

Yes, maybe. But I'm leaning towards converting everything into DNF form, it feels natural for expressions we would be writing. We can try and choose later, when we will have more material to consider.

**Comment by matejak** [comment]
> ...
> > 2. There are other files included - is it about dealing with version numbers?
> > 
> 
> 
> Yes, the `packaging` module is related to version specifiers.

What about setuptools? Its [pkg_resources](https://setuptools.pypa.io/en/latest/pkg_resources.html#parsing-utilities) package contains functionality implementing PEP 440, and setuptools are widely available.

**Comment by matejak** [comment]
> > ```
> > 1. The result of parsing the expression is either `Function` or `Symbol`. `Function` is basically a graph node, its `args` are other nodes or leaf nodes. Therefore, the representation of a boolean expression is available in form of a tree.
> > ```
> 
> 
> Yes.

It would be more user-friendly if a result of parsing is `OperatorNode` instead of `Function` with children under `child_nodes` instead of `args`,  and `LeafNode` instead of a `Symbol`. Once the expression is parsed, there is no node, just the tree, and especially the `args` attribute to access children nodes is confusing.

**Comment by evgenyz** [comment]
> It would be more user-friendly if a result of parsing is `OperatorNode` instead of `Function` with children under `child_nodes` instead of `args`, and `LeafNode` instead of a `Symbol`. Once the expression is parsed, there is no node, just the tree, and especially the `args` attribute to access children nodes is confusing.

Well, first of all all of this is inherited from the `boolean.py`. We, of course, could create proxy classes and methods to do the renaming, but, honestly, I doubt that it worth the effort.

**Comment by evgenyz** [comment]
> What about setuptools? Its [pkg_resources](https://setuptools.pypa.io/en/latest/pkg_resources.html#parsing-utilities) package contains functionality implementing PEP 440, and setuptools are widely available.

Actually, we can use `pkg_resources`. During the first round of matching I overlooked the fact that there is a requirement parser (PEP 508) ready to be used: https://setuptools.pypa.io/en/latest/pkg_resources.html#requirements-parsing. Thanks for the hint for redoing my once-over. I'll update the PR.

**Comment by vojtapolasek** [comment]
> > > ```
> > > 1. The result of parsing the expression is either `Function` or `Symbol`. `Function` is basically a graph node, its `args` are other nodes or leaf nodes. Therefore, the representation of a boolean expression is available in form of a tree.
> > > ```
> > 
> > 
> > Yes.
> 
> It would be more user-friendly if a result of parsing is `OperatorNode` instead of `Function` with children under `child_nodes` instead of `args`, and `LeafNode` instead of a `Symbol`. Once the expression is parsed, there is no node, just the tree, and especially the `args` attribute to access children nodes is confusing.

I think it is not necessary. These "Symbol" and "Function" classes will be anyway subclassed by CPEALLogicalTest and CPEALFactRef, at least that is my solution so far. So the technical details will be hidden anyway. Do you feel strongly about having it more user friendly?

**Review by vojtapolasek — APPROVED** [review]
Thank you very much for this implementation. I like it. I managed to incorporate it into the build system in my private branch and I did not find any serious obstacles.
I think this can be merged as soon as all questions get answered.

**Review by yuumasato — APPROVED** [review]
LGTM.

I understand @matejak's concern but also agree that creating proxy classes may not be necessary.
@vojtapolasek has some implementation using this code and we will see there if it will bee too confusing with the classes as they are.

**Pull request merged**